### PR TITLE
Reducing closure and list allocations in LambdaCompiler

### PR DESCRIPTION
**Pull request by bartdesmet**
This addresses issue #11323 by:

* Changing `WriteBack` delegate to take in `LambdaCompiler` so the closure doesn't have to include `this`
* Splitting `AddressOfWriteBack` methods by adding a `Core` method to avoid closure allocation
* Reducing closure size by reduce the number of locals needed in the write-back
* Removing cleverness in `EmitArray` where a callback was used by inlining the logic in its single call site
* Lazily allocating `List<WriteBack>` at the point it's needed, using `null` to represent no write-backs
* Using `List<WriteBack>` instead of `IList<WriteBack>` for `foreach` so it binds to struct-based `GetEnumerator`

CC @VSadov

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs [hunk 1]
```diff
@@ -1080,26 +1080,16 @@ internal static void EmitArray<T>(this ILGenerator il, IList<T> items)
...
-        /// Emits an array of values of count size.  The items are emitted via the callback
-        /// which is provided with the current item index to emit.
+        /// Emits an array of values of count size.
         /// </summary>
-        internal static void EmitArray(this ILGenerator il, Type elementType, int count, Action<int> emit)
+        internal static void EmitArray(this ILGenerator il, Type elementType, int count)
         {
             ContractUtils.RequiresNotNull(elementType, nameof(elementType));
-            ContractUtils.RequiresNotNull(emit, nameof(emit));
+
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs]: Using issue #11325 to track changes from `ContractUtils` to `Debug` asserts where possible. Will make separate PR for this.

**Comment by bartdesmet** [comment]
Test Innerloop Windows_NT Debug Build and Test please (ran into one of the issues tracked in #9228)

**Comment by VSadov** [comment]
LGTM. Thanks!!
Need to resolve merge conflicts

**Comment by bartdesmet** [comment]
Test Innerloop OSX Debug Build and Test please (Build machine went offline during the build.)
